File: client-libs/luncho-typescript-aurelia/luncho.ts
/**
   Fast Luncho API by caching. Don't this
*/
import { LunchoData } from './models';
import { LunchoApi, ILunchoDataParams } from './LunchoApi';

export type CountryCode = string;

export class Luncho extends LunchoApi {

    lunchoDataMap: { [key: string]: LunchoData} = {};  // Cache {CountryCode: LunchoData}
    allLunchoDatasFetched = false;
    countryMap: { [key: string]: string; };


    /**
       Returns a local data for the given country code using cache.
    */
    async lunchoData(param: ILunchoDataParams ): Promise<LunchoData> {
        const lunchoData: LunchoData = this.lunchoDataMap[param.countryCode];
        if (lunchoData) {
            return Promise.resolve(lunchoData);
        }

        return super.lunchoData(param)
            .then((lunchoData: LunchoData) => {
                this.lunchoDataMap[param.countryCode] = lunchoData;
                  return(lunchoData);
            });
    }

    /**
       Returns a local data for the given country code using cache.
    */
    async lunchoDatas(): Promise<{ [key: string]: LunchoData} > {
        if (this.allLunchoDatasFetched) {
            return Promise.resolve(this.lunchoDataMap);
        }

        return super.lunchoDatas()
            .then((lunchoDatas: { [key: string]: LunchoData}) => {
                this.lunchoDataMap = lunchoDatas;
                this.allLunchoDatasFetched = true;
                return(lunchoDatas);
            });
    }

    /**
       Returns a local data for the given country code using cache.
    */
    async getCountries(): Promise<{ [key: string]: string; }> {
        if (this.countryMap) {
            return Promise.resolve(this.countryMap);
        }

        return super.countries()
            .then((countryMap: { [key: string]: string; }) => {
                this.countryMap = countryMap;
                  return(countryMap);
            });
    }

    /**
       Returns a local currency value from the given Luncho value using cache.
    */
    async localCurrencyFromLuncho(lunchoValue: number, countryCode?: string): Promise<number> {
        return this.lunchoData({countryCode: countryCode})
            .then((lunchoData: LunchoData) => {
                  return(lunchoData.dollar_per_luncho * lunchoData.ppp * lunchoValue);
            });
    }

    /**
       Returns a US Dollar value from the given Luncho value using cache.
    */
    async USDollarFromLuncho(lunchoValue: number, countryCode?: string): Promise<number> {
        return this.lunchoData({countryCode: countryCode})
            .then((lunchoData: LunchoData) => {
                const local_currency_value = lunchoData.dollar_per_luncho * lunchoData.ppp * lunchoValue;
                return(local_currency_value / lunchoData.exchange_rate);
            });
    }
}

File: client-libs/luncho_python/luncho_python/api/luncho.py
'''
   Fast Luncho API client by caching. Use this and don't this LunchoApi.py

  @author: HIRANO Satoshi
  @date: 2021-5-18
'''
import time
from typing import List, Dict, Tuple, Callable, Union, Any, Set, ClassVar, Type, Optional, cast

from luncho_python.api.luncho_api import LunchoApi
from luncho_python.api_client import ApiClient, Endpoint as _Endpoint
from luncho_python.model.luncho_data import LunchoData

CountryCode = str

class Luncho(LunchoApi):
    def __init__(self, api_client=None):
        self.lunchoDataMap: Dict[CountryCode, LunchoData] = {}  # Cache {CountryCode: LunchoData}
        self.allLunchoDatasFetched: bool = False
        self.countryMap: Dict[CountryCode, str] = {}       # { CountryCode: name }
        super().__init__(api_client)


    def luncho_data_fast(self, country_code: CountryCode, **kwargs) -> LunchoData:
        '''
           Returns a Luncho data for the given country code using cache.
        '''
        lunchoData: Optional[LunchoData] = self.lunchoDataMap.get(country_code)
        if lunchoData and lunchoData.expiration > time.time():
            return lunchoData

        lunchoData = self.luncho_data(country_code, **kwargs)
        lunchoData = cast(LunchoData, lunchoData)
        self.lunchoDataMap[country_code] = lunchoData
        return lunchoData

    def luncho_datas_fast(self, **kwargs) -> Dict[CountryCode, LunchoData]:
        '''
        Returns a local data for the given country code using cache.
        '''
        if self.allLunchoDatasFetched:
            lunchoData: Optional[LunchoData] = self.lunchoDataMap.get('JP')
            if lunchoData and lunchoData.expiration > time.time():
                return self.lunchoDataMap

        lunchoDatas: Dict[CountryCode, LunchoData] = self.luncho_datas(**kwargs)
        self.lunchoDataMap = lunchoDatas
        self.allLunchoDatasFetched = True
        return lunchoDatas


    def countries_fast(self, **kwargs) -> Dict[CountryCode, str]:
        '''
        Returns a local data for the given country code using cache.
        '''

        #import pdb; pdb.set_trace()
        if self.countryMap:
            return self.countryMap

        self.countryMap = self.countries(**kwargs)
        return self.countryMap

    def localCurrencyFromLuncho(self, lunchoValue: float, countryCode: str, **kwargs) -> float:
        '''
        Returns a local currency value from the given Luncho value using cache.
        '''

        lunchoData: LunchoData = self.luncho_data_fast(countryCode, **kwargs)
        return lunchoData.dollar_per_luncho * lunchoData.ppp * lunchoValue

    def USDollarFromLuncho(self, lunchoValue: float, countryCode: str, **kwargs) -> float:
        '''
        Returns a US Dollar value from the given Luncho value using cache.
        '''

        lunchoData: LunchoData = self.luncho_data_fast(countryCode, **kwargs)
        if lunchoData.exchange_rate > 0:
            local_currency_value: float = lunchoData.dollar_per_luncho * lunchoData.ppp * lunchoValue;
            return local_currency_value / lunchoData.exchange_rate
        else:
            return 0.0
